README
What is this?
-------------
It's a really small but useful chat app for a not more than a handful of users (because there are no rooms).
The focus lies on smallness and network recovery, i.e. if the network is interrupted, there should be no "lost" messages but they should be sent/received when the network is available again. Also messages are by default encrypted (with a predefined key, which should really be changed when running the clients).

Old messages (how far the history goes back can be changed) are sent to newly connected clients, so even if they arrive late in a conversation, they still get to see how it developed.

Note that clients and server exchange version information in the initial handshake, so they have to match.

The whole system consists of a server process on one computer and each user running a client process on their machine. The clients can be graphical user interfaces or text/console based.

How does it work?
-----------------
The clients regularly poll the server for new messages and apart from that send heartbeats to signify that they are still running. The messages are then shown in the graphical user interface and a popup is shown in the notification tray (e.g. in Windows). Clicking on the little tapir icon in the tray will bring the conversation window to the front.
The conversation window's upper text field shows (in faded colours for unobtrusiveness) who is currently online.
Whenever the text input field (bottom field in conversation window) is green, the client is online (i.e. it's connected to the server and both know of each other). When it's grey, the network is interrupted or the server is down.

How does the client program work?
---------------------------------
The GUI (graphical user interface) is divided into three areas:
- the top field, which contains a list of users that are online just now (as the server sees them)
- the main field in the middle, which contains the list of messages that the client received
- the bottom field (green or gray, according to online status), into which messages can be written to be sent to all other users

A notification tray popup is shown each time a new message comes in (this behaviour can be turned off by a command line option).
Clicking on the tray icon or popup will bring the converation window to the top.

How to write a message:
-----------------------
Write some text into the green (or gray, if offline).

What is needed to run this:
---------------------------
- A publicly (or at least on the local network if all clients are local) available computer to run the server in.
- The Java Runtime Environment version 1.6 or later on server and client machines.

How to run it:
--------------
(N.B.: change the name of the jar file according to the file downloaded or built)

Easiest way to run the server:
java -jar tc-20151210-server.jar

Easiest way to run the client:
java -jar tc-20151210-client.jar -u <username> 

To print all available options:
java -jar tc-20151210-client.jar  -help
or
java -jar tc-20151210-server.jar  -help

Available options:
  -help this message
  -g <log file>   (should be first command line option; '-' for stdout (default))
  -j <journal file>   (only for Hub; logs all user messages)
  -l [FINEST|FINER|FINE|INFO|WARNING|SEVERE]  -> log level, should be first or second option
  -h <server host>    (only for TextUI or SwingUI)
  -p <server port, default 64321
  -u <user name>      (only for TextUI or SwingUI)
  -r <local port>     (only for TextUI or SwingUI)
  -c | -noc   -> use|disable encryption
  -pc | -nopc -> use|disable payload encryption
  -nopo       -> disable popup balloons
  -i <ip address[,ip address]>  -> allowed ip addresses
  -a <minutes>   -> how many minutes' worth of archive to request (default 30)
  -b <seconds>   -> heartbeat interval (default 30 seconds)
  -k <secret key>  -> 16 bytes in the format -33,-59,-1,106,... ; values: -128 .. 127
  -fs <font size>  -> font size, can be float; default: 11.0; must come before -fo
  -fo <font>       -> font; must be full name; list all fonts with -l FINEST -fo x
Call as: java -cp tc.jar tc.Hub     -> Server mode
         java -cp tc.jar tc.TextUI  -> text interface
         java -cp tc.jar tc.SwingUI -> graphical interface
         java -jar tc.jar           -> same as tc.Hub
         java -Djava.awt.headless=true  -> if no System Tray notification wished
         java -cp tc.jar tc.Hub -i 192.168,10.,172.16.  -> only accept local ip addresses
         java -jar tc-20151210-client.jar -u user2 -fs 16 -fo Calibri -> Use "Calibri" font in size 16


How to change the secret key?
-----------------------------
It's a command line option that must be supplied for both server and clients, in a comma separated format, e.g.:
-k "1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6"

So a complete start command looks like:
java -jar tc-20151210-server.jar -k "1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6"
And a couple of client:
java -jar tc-20151210-client.jar -u Alice -k "1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6"
java -jar tc-20151210-client.jar -u Bob -k "1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6"


How to build it:
----------------
There is a really small build script (works with the Bash shell in *nix and Cygwin under Windows). The final jar files are copied into the "jars" directory.

Why is it called TapirChat?
---------------------------
Because tapirs are nice creatures and the tray icon is a little tapir (kudos to www.tapirsoft.de who have given public permission to use their logo).

"Secret" commands:
------------------
A few special words written into the text input field will be interpreted by the client as special commands:
"WHO" - update the topmost text field with who else is online
"LOG FINEST" - sets the log level to "finest"; other values are: "FINER", "FINE", "CONFIG", "INFO", "WARNING", "SEVERE"
"STATS" - prints number of sent/received messages (includes heartbeat messages) and amount of data transferred to the console window.



Changelog
2012-08-07
	main(): grauer Hintergrund, wenn Server nicht erreichbar. Grund: connect()
	war innerhalb invokeLater().
	Loesung: SwingUI-initialisierung in Event Handling Thread, conneect()
	ausserhalb nach invokeAndWait()

2012-08-01
	run setServerNotResponding on Event Dispatch Thread
	
2012-07-30
	invoke conversationDoc changes on Event Dispatch Thread
	invoke new SwingUI on Event Dispatch Thread

2012-07-25
	Mehr logging in setServerNotResponding

2012-02-27
	Fonts: print default font name (-l CONFIG)

2011-08-23
	List all fonts with -l FINEST -fo x
	New option "-fs" for font size
	Potential NPE in UserInputThread after getUserInput(): now remove invalid
	chars *after* testing for null

2011-08-22
	Fix TextUI: filter out NULL und "|" (Message.FS) direkt in
	UserInputThread.

2011-08-09
	notificationField read-only

2011-08-08
	Versionsnummer zurueck auf tc0.3 (keine Protokollaenderung fuer Hub
		noetig)
	gesendete/empfangene Anzahl bytes ausgeben mit -l FINEST
	disable tray with -nopop
	usage Hilfe mit "-help"
	SECRET_DATA: "DATA" gibt sent/received aus in log file/Konsole

2011-08-05
	Nachricht verschlüsseln (aber nicht id/timestamp etc) mit ObjectCrypter;
	New version tc0.4

2011-07-12
	Fix bug: Hub bricht ab bei unbekanntem Message_type (enum parse error)
	Fix findbugs: Hub,SwingUI,Logger - DateFormat as ThreadLocal (Df. is not thread safe)
	Fix Options: more error message if argument missing or unknown argument;
		let caller decide what to do on Options.unknownOptionFound==true
	Fix TextUI: ignore message queue full
	WHO status leeren, wenn offline
	status bar: "Offline seit ..."


2011-07-08
	WHO ausgegraut dargestellt in SwingUI.writeConversationMessage()

2011-07-01
	WHO direkt nach Programmstart (1 Sek. delay statt acktime/2)

2011-06-30
	Notification pane at top to show who is online (regularly updated with
	same interval as ACK)

2011-03-22
	Options: use CONFIG log level to echo settings

2011-03-18
	TextUI: changed system.exit calls to "continue" in connect()

2011-03-16
	TextUI: sleep for heartbeatInterval on connect exception instead of
	hard wired 30 secs

2011-03-11
	Tray: set offline after creating
	Logger: flush after printing "log file opened" line

2011-03-10
	Hub/Options: implement journal file which captures every user message
		also for reading in on startup
	Logger: fix dolog(Level,String,Exception)
	Message types as Enum
	Allow log level to be adjusted after starting with "secret" "LOG" command

2011-03-09
	TextUI: fixed bug, compare MSG_TYPE_HEARTBEAT with msgBody instead of
		msgType when inserting into sentMessageIds
	TextUI: sentMessageIds, receivedMessageIds as LRU caches (potential memory
		leak)
	INFO: encryption may have a memory leak

2011-03-08
	Hub,Options: moved default PORT to Options
	TextUI: removed reference to Hub.PORT
	TextUI: sentMessageIds: only include non-heartbeat ids
	TextUI: only change tray status if there actually was a change in
		serverNotResponding (probably memory leak in Tray.setImageIcon)

2011-03-07
	Hub: VERSION to 0.3
	Options: heartbeatInterval (-b)
	Client/TextUI: support heartbeat interval message in handshake

2011-03-01
	Message: Use java.lang.UUID to create message IDs
	Hub: VERSION upped to 0.2
	Fix NumberFormatException if Message from structuredString has timestamp
	in non-Long format

2011-02-28
	Client:288 don't repeat message id on returning WHOSONLINE, otherwise
	TextUI thinks it's an ACK message

2011-02-24
	Reset tray item on clicking into conversation or input area

2011-02-18
	Bugfixes (findbugs)
	Start UserInputThread from connect(), not constructor

2011-02-16
	Type safe Options.
	User-definable encryption keys "-k" option
	Bug: -Djava.awt.headless=true caused NullPointerException in
		Tray.setOnline()/setOffline()

2011-02-15
	Fix Logger: date field size 18 instead of 16

2011-02-14
	TextUI: setServerNotResponding(false) in connect after handshake complete
	Hub, SwingUI, Client, Message: preparations for ISTYPING messages


src/tc/Options.java
package tc;

import java.awt.Font;
import java.awt.GraphicsEnvironment;
import java.io.File;
import java.io.IOException;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.logging.Level;

// if options logging is to be shown, "-l FINE" must be the FIRST option!
public final class Options {
    public static final byte EOL = 0;

    public Integer optPort = 64321;
    public String optHost;
    public String optUser;
    public String[] optIpAddresses;
    public Integer optLocalPort = 0;
    public Boolean optUseEncryption = true;
    public Boolean optUsePayloadEncryption = false;
    public Integer optArchiveMinutes = 720;
    public Font optFont;
    public float optFontSize = 11.0f;
    public File optLogfile;
    public File optJournalFile;
    public Boolean optPopups = true;
    public byte[] optSecretKey;
    public Integer optHeartbeat = 30*1000;
    public boolean unknownOptionFound = false;

    public Options (String[] args) {
        Deque<String> deq = new ArrayDeque<String>(Arrays.asList(args));
        
        while(!deq.isEmpty()){
            String opt = deq.pop();
            if (opt.startsWith("-")){
                
                // First the options without argument
                if (opt.startsWith("-c")) {
                    optUseEncryption = true;
                    Logger.log (Level.CONFIG, "Enabling encryption.");
                } else if (opt.startsWith("-help")) {
                    printUsage();
                    System.exit(0);
                } else if (opt.startsWith("-noc")) {
                    optUseEncryption = false;
                    Logger.log (Level.CONFIG, "Disabling encryption.");
                } else if (opt.startsWith("-pc")) {
                    optUsePayloadEncryption = true;
                    Logger.log (Level.CONFIG, "Enabling payload encryption.");
                } else if (opt.startsWith("-nopc")) {
                    optUsePayloadEncryption = false;
                    Logger.log (Level.CONFIG, "Disabling payload encryption.");
                }  else if (opt.startsWith("-nopo")) {
                    optPopups = false;
                    Logger.log (Level.CONFIG, "Disabling popups");
                } else {
                    
                    // Now options with argument
                    if (deq.isEmpty()) {
                        // second part of -x <option> missing
                        Logger.log (Level.WARNING, "Expected 1 argument for : " + opt);
                        unknownOptionFound = true;
                        break;
                    }
                
                    String parm = deq.poll();
                    // doesn't work with -k: argument list may start with negative number
                    if (opt.startsWith("-k") && parm.startsWith("-")) {
                        Logger.log(Level.SEVERE, "Argument for "+opt+" must not start with '-': " + parm);
                        unknownOptionFound = true;
                        break;
                    }
                    if (opt.startsWith("-p")){ // Port
                        try {
                            optPort = Integer.parseInt(parm);
                        } catch (NumberFormatException nfe){
                            Logger.log(Level.SEVERE, "Could not parse port "+parm);
                            throw new RuntimeException("Could not parse port "+parm);
                        }
                    } else if (opt.startsWith("-h")){ // Receiver Host
                        optHost = parm;
                    } else if (opt.startsWith("-u")){ // user
                        optUser = parm;
                    } else if (opt.startsWith("-i")){ // allowed ip addresses that can connect
                        optIpAddresses = parm.split(",");
                    } else if (opt.startsWith("-l")){ // log level
                        if (parm.startsWith("FINEST")){
                            Logger.setLogLevel(Level.FINEST);
                        } else if (parm.startsWith("FINER")){
                            Logger.setLogLevel(Level.FINER);
                        } else if (parm.startsWith("FINE")){
                            Logger.setLogLevel(Level.FINE);
                        } else if (parm.startsWith("CONFIG")){
                            Logger.setLogLevel(Level.CONFIG);
                        } else if (parm.startsWith("INFO")){
                            Logger.setLogLevel(Level.INFO);
                        } else if (parm.startsWith("WARNING")){
                            Logger.setLogLevel(Level.WARNING);
                        } else if (parm.startsWith("SEVERE")){
                            Logger.setLogLevel(Level.SEVERE);
                        } else {
                            Logger.log (Level.WARNING, "Unknown log level: " + parm);
                        }
                        Logger.log (Level.CONFIG, "Setting log level to " + Logger.getLevel());
                    } else if (opt.startsWith("-r")) {
                        try {
                            optLocalPort = Integer.parseInt(parm);
                            Logger.log (Level.CONFIG, "Setting local port to " + parm);
                        } catch (NumberFormatException nfe){
                            Logger.log(Level.SEVERE, "Could not parse local port "+parm);
                            throw new RuntimeException("Could not parse local port "+parm);
                        }
                    }  else if (opt.startsWith("-a")) {
                        try {
                            optArchiveMinutes = Integer.parseInt(parm);
                        } catch (NumberFormatException nfe){
                            Logger.log(Level.SEVERE, "Could not parse archive minutes "+parm);
                            throw new RuntimeException("Could not parse archive minutes "+parm);
                        }
                        Logger.log (Level.CONFIG, "Setting number of minutes of archive to request to " + parm);
                    } else if (opt.startsWith("-b")) {
                        try {
                            optHeartbeat = Integer.parseInt(parm) * 1000;
                        } catch (NumberFormatException nfe){
                            Logger.log(Level.SEVERE, "Could not parse heartbeat interval "+parm);
                            throw new RuntimeException("Could not parse heartbeat interval "+parm);
                        }
                        Logger.log (Level.CONFIG, "Setting heartbeat interval to " + parm + " seconds");
                    } else if (opt.startsWith("-fo")) {
                        GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
                        String fontName = parm;
                        if (ge != null){
                            try {
                                Font[] allFonts = ge.getAllFonts();
                                for (Font f : allFonts){
                                    Logger.log(Level.FINEST, "Font on system: "+f.getName());
                                    if (fontName.equals(f.getName())){
                                        optFont = f.deriveFont(optFontSize);
                                        Logger.log(Level.CONFIG, "Found font '" + optFont.getName() + "'");
                                        break;
                                    }
                                }
                            } catch (NullPointerException npe) {
                            }
                        }
                        if (optFont == null) {
                            Logger.log(Level.WARNING, "Could not find font " + fontName + " on this machine");
                        } else {
                            Logger.log (Level.CONFIG, "Setting font to " + parm);
                        }
                    } else if (opt.startsWith("-fs")) {
                        try {
                            optFontSize = Float.parseFloat(parm);
                        } catch (NumberFormatException nfe) {
                            Logger.log(Level.WARNING, "Could not parse font size: " + parm);
                            break;
                        }
                        Logger.log (Level.CONFIG, "Setting font size to " + optFontSize );
                    } else if (opt.startsWith("-k")) {
                        // 16 bytes -128 .. 127
                        String[] bytes = parm.split(",");
                        if (bytes.length != 16) {
                            Logger.log(Level.SEVERE, "Expected 16 bytes for secret key");
                            unknownOptionFound = true;
                            break;
                        }
                        int i = 0;
                        optSecretKey = new byte[16];
                        try {
                            for (String s : bytes){
                                optSecretKey[i] = Byte.parseByte(s);
                                i++;
                            }
                        } catch (NumberFormatException nfe) {
                            Logger.log(Level.SEVERE, "Invalid byte value: " + bytes[i]);
                            unknownOptionFound = true;
                            break;
                        }
                        Logger.log (Level.CONFIG, "Setting secret key to " + parm );
                    } else if (opt.startsWith("-g")) {
                        optLogfile = new File(parm);
                        Logger.log (Level.CONFIG, "Setting log file to " + parm);
                        try {
                            Logger.setLogFile(optLogfile);
                        } catch (IOException e) {
                            System.err.println("Could not open log file for writing: " + parm);
                            optLogfile = null;
                        }
                    } else if (opt.startsWith("-j")) {
                       optJournalFile = new File(parm);
                       Logger.log (Level.CONFIG, "Setting journal file to " + parm);
                    } else {
                        Logger.log (Level.WARNING, "Unknown option: " + opt);
                        unknownOptionFound = true;
                    }
                }
            } else { 
                Logger.log (Level.WARNING, "Unknown option: " + opt);
                unknownOptionFound = true;
                break;
                    
            }
        }

    }

    public void printUsage(){
        System.out.println("Available options:");
        System.out.println("  -help this message");
        System.out.println("  -g <log file>   (should be first command line option; '-' for stdout (default))");
        System.out.println("  -j <journal file>   (only for Hub; logs all user messages)");
        System.out.println("  -l [FINEST|FINER|FINE|INFO|WARNING|SEVERE]  -> log level, should be first or second option");
        System.out.println("  -h <server host>    (only for TextUI or SwingUI)");
        System.out.println("  -p <server port, default " + optPort);
        System.out.println("  -u <user name>      (only for TextUI or SwingUI)");
        System.out.println("  -r <local port>     (only for TextUI or SwingUI)");
        System.out.println("  -c | -noc   -> use|disable encryption");
        System.out.println("  -pc | -nopc -> use|disable payload encryption");
        System.out.println("  -nopo       -> disable popup balloons");
        System.out.println("  -i <ip address[,ip address]>  -> allowed ip addresses");
        System.out.println("  -a <minutes>   -> how many minutes' worth of archive to request (default 30)");
        System.out.println("  -b <seconds>   -> heartbeat interval (default 30 seconds)");
        System.out.println("  -k <secret key>  -> 16 bytes in the format -33,-59,-1,106,... ; values: -128 .. 127");
        System.out.println("  -fs <font size>  -> font size, can be float; default: 11.0; must come before -fo");
        System.out.println("  -fo <font>       -> font; must be full name; list all fonts with -l FINEST -fo x");
        System.out.println("Call as: java -cp tc.jar tc.Hub     -> Server mode");
        System.out.println("         java -cp tc.jar tc.TextUI  -> text interface");
        System.out.println("         java -cp tc.jar tc.SwingUI -> graphical interface");
        System.out.println("         java -jar tc.jar           -> same as tc.Hub");
        System.out.println("         java -Djava.awt.headless=true  -> if no System Tray notification wished");
        System.out.println("         java -cp tc.jar tc.Hub -i 192.168,10.,172.16.  -> only accept local ip addresses");
        
    }

}
